### PR TITLE
LibJS: A bunch of parser fixes

### LINKED ISSUE #11078: LibJS: Unable to use keywords as property names in object literals
For example: `{ url: "", async: false }` produces:
```
{ url: "", async: false }
                ^
Uncaught exception: [SyntaxError] Unexpected token Colon. Expected Semicolon (line: 1, column: 17)
```

Similarly, `{ url: "", function: false }` produces:
```
{ url: "", function: false }
                   ^
Uncaught exception: [SyntaxError] Unexpected token Colon. Expected ParenOpen (line: 1, column: 20)
```

This affects jQuery for example: https://github.com/jquery/jquery/blob/a684e6ba836f7c553968d7d026ed7941e1a612d8/src/ajax.js#L304

### PR DESCRIPTION
**Pull request by davidot**
Fixes #11078 
And much more

~Note this is incompatible with #11041 as is because I renamed a variable so one will need to be fixed up.~ Already merged

### CODE REVIEW

File: Userland/Libraries/LibJS/Parser.cpp [hunk 1]
```diff
@@ -1230,8 +1232,8 @@ Parser::PrimaryExpressionParseResult Parser::parse_primary_expression()
             syntax_error("'super' keyword unexpected here");
         return { create_ast_node<SuperExpression>({ m_state.current_token.filename(), rule_start.position(), position() }) };
     case TokenType::EscapedKeyword:
-        if (m_state.strict_mode || (m_state.current_token.value() != "let"sv && (m_state.in_generator_function_context || m_state.current_token.value() != "yield"sv) && (m_state.in_async_function_context || m_state.current_token.value() != "await"sv)))
-            syntax_error("Keyword must not contain escaped characters");
+        if (match_invalid_escaped_keyword())
+            syntax_error("Keyword must not contain escaped characters 1");
```
Comment (IdanHo) [inline Userland/Libraries/LibJS/Parser.cpp:1351]: typo?

Comment (davidot) [reply Userland/Libraries/LibJS/Parser.cpp]: Yeah forgot to remove that debugging

File: Userland/Libraries/LibJS/Parser.cpp [hunk 2]
```diff
@@ -597,7 +597,7 @@ bool Parser::match_invalid_escaped_keyword() const
         return false;
     auto token_value = m_state.current_token.value();
     if (token_value == "await"sv)
-        return m_program_type == Program::Type::Module || m_state.in_async_function_context;
+        return m_program_type == Program::Type::Module || m_state.await_expression_is_valid_context;
```
Comment (alimpfard) [inline Userland/Libraries/LibJS/Parser.cpp:655]: `await_expression_is_valid_context` doesn't sound very nice, how about just what the commit message suggests: `await_expression_is_valid`?

File: Userland/Libraries/LibJS/Parser.cpp [hunk 3]
```diff
@@ -1432,7 +1432,8 @@ Parser::PrimaryExpressionParseResult Parser::parse_primary_expression()
             goto read_as_identifier;
         return { parse_await_expression() };
     case TokenType::PrivateIdentifier:
-        VERIFY(next_token().type() == TokenType::In);
+        if (next_token().type() != TokenType::In)
+            syntax_error("Cannot have a private identifier in expression if not followed by 'in'");
```
Comment (alimpfard) [inline Userland/Libraries/LibJS/Parser.cpp:1436]: (not an issue) TIL that this is a thing, interesting...

**Review by linusg — CHANGES_REQUESTED** [review]
Please add more tests. Ideally every fix should be accompanied by a test, and for most of these it can be a single line. Having tests allows the reviewer to verify that the implementation is correct more easily, and will prevent us from regressing anything that was once known to work and is tested.

I'm also not a fan of vague commit messages such as:

> LibJS: Allow some instances of escaped 'async' identifiers
> LibJS: Disallow await keywords in static init blocks in more cases 

"some", "in more cases"? Since you didn't go into detail in the commit message body in either of these, at the very least add some tests to show what is now considered (in)valid that previously wasn't.

File: Userland/Libraries/LibJS/Parser.cpp (hunk 2 above)
Comment (davidot) [reply Userland/Libraries/LibJS/Parser.cpp:655]: Yeah I'm still not super happy with the name, ideally I would like `+await` or something as that is how it is used in the spec. Anyway I'm sure someone will figure out a nice name later.

**Comment by davidot** [comment]
I added test to all the commits (except ones about modules because we don't really have a way to tests those) and attempted to clarify the commit messages.
Unfortunately these are just edge cases of js parsing so I'm not sure what else to write then just explain a little bit.

File: Userland/Libraries/LibJS/Tests/test-common.js [hunk 4]
```diff
@@ -353,7 +353,12 @@ class ExpectationError extends Error {
         toEval() {
             this.__expect(typeof this.target === "string");
             const success = canParseSource(this.target);
-            this.__expect(this.inverted ? !success : success);
+            this.__expect(
+                this.inverted ? !success : success,
+                () =>
+                    `Expected _${valueToString(this.target)}_` +
+                    (this.inverted ? "not to eval but it did" : "to eval but it didn't")
+            );
```
Comment (davidot) [inline Userland/Libraries/LibJS/Tests/test-common.js:361]: While writing the tests I hit these a lot, should this change also be tested? It is just human friendly output

Comment (linusg) [reply Userland/Libraries/LibJS/Tests/test-common.js:361]: nope, that's ok. Thanks for making them more useful though!

File: Userland/Libraries/LibJS/Tests/syntax/async-await.js [hunk 5]
```diff
@@ -167,6 +168,16 @@ describe("non async function declaration usage of async still works", () => {
         const evalResult = eval("async >= 2");
         expect(evalResult).toBeTrue();
     });
+
+    test("async with line ending does not create a function", () => {
+        expect(() => {
+            async;
+            function f() {}
```
Comment (linusg) [inline Userland/Libraries/LibJS/Tests/syntax/async-await.js:175]: This was probably "fixed" by prettier, should not have the semicolon

**Review by linusg — APPROVED** [review]
Perfect, thanks for adding all those tests. Great work!

File: Userland/Libraries/LibJS/Tests/syntax/async-await.js (hunk 5 above)
Comment (davidot) [reply Userland/Libraries/LibJS/Tests/syntax/async-await.js]: Added prettier ignore with a comment
